homeassistant/components/heos/coordinator.py

```diff
@@ -0,0 +1,93 @@
+"""HEOS integration coordinator.
+
+Control of all HEOS devices is through connection to a single device. Data is pushed through events.
+The coordinator is responsible for refreshing data in response to system-wide events and notifying
+entities to update. Entities subscribe to entity-specific updates within the entity class itself.
+"""
+
+import logging
+
+from pyheos import Credentials, Heos, HeosError, HeosOptions, MediaItem
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from . import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HeosCoordinator(DataUpdateCoordinator[None]):
+    """Define the HEOS integration coordinator."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Set up the coordinator and set in config_entry."""
+        self.host: str = config_entry.data[CONF_HOST]
+        credentials: Credentials | None = None
+        if config_entry.options:
+            credentials = Credentials(
+                config_entry.options[CONF_USERNAME], config_entry.options[CONF_PASSWORD]
+            )
+        # Setting all_progress_events=False ensures that we only receive a
+        # media position update upon start of playback or when media changes
+        self.heos = Heos(
+            HeosOptions(
+                self.host,
+                all_progress_events=False,
+                auto_reconnect=True,
+                credentials=credentials,
+            )
+        )
+        self.favorites: dict[int, MediaItem] = {}
+        self.inputs: list[MediaItem] = []
+        super().__init__(hass, _LOGGER, config_entry=config_entry, name=DOMAIN)
+
+    async def async_setup(self) -> None:
+        """Set up the coordinator; connect to the host; and retrieve initial data."""
+        # Add before connect as it may occur during initial connection
+        self.heos.add_on_user_credentials_invalid(self._async_on_auth_failure)
+        # Connect to the device
+        try:
+            await self.heos.connect()
+        except HeosError as error:
+            raise ConfigEntryNotReady from error
```

Comment:
In the future, this (and the next try/except block) will have a specific message informing the user why setup failed. I can add that to this PR if that helps justify the repeated try/except blocks (I was trying to keep this PR as small as possible initially.)